TestModels/Constraints/runtimes/rust/src/types/error.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+// Do not modify this file. This file is machine generated, and any changes to it will be overwritten.
+#[derive(::std::clone::Clone, ::std::fmt::Debug, ::std::cmp::PartialEq)]
+pub enum Error {
+    SimpleConstraintsException {
+    message: ::std::string::String,
+},
+    CollectionOfErrors {
+        list: ::std::vec::Vec<Self>,
+        message: ::std::string::String,
+    },
+    ValidationError(ValidationError),
```

Comment:
Could this be `ValidationError(BuildError)` instead?

Reply:
For implementing constraint trait validation only, yes, I believe this could just directly wrap a `BuildError`.

There's two levels of indirection here:

- This PR uses an  `Rc<dyn std::error::Error>` instead of `BuildError` directly, because I was not (and am not) sure that all validation errors will be `BuildError`s.
- In order to safely up-/down-cast the contents of the `Error::ValidationError` variant from the `dyn Any` in an arbitrary `dafny_runtime::Object`, it's best to define a unique type that we own - that is, the `ValidationError` struct (not the enum variant). If we defined `Enum::ValidationError(Rc<...>)` or `Enum::ValidationError<BuildError>`, then it seems possible for us to mishandle a `dafny_runtime::Object` containing an `Rc<?>` or `BuildError` from somewhere else.

In my mind, this doesn't make the library harder to use: the typical library user would mainly care whether or not the error variant is a `ValidationError` or something else. If they want the message or the underlying `dyn std::error::Error`, there's no need to manually traverse the indirections: they can use `to_string()` and `source()` as you would for any other `std::error::Error`.

Reply:
Fair points. What would it look like for an end-user to react differently to a validation error, and actually dig out the offending field from the wrapped `BuildError`?

Reply:
There isn't a clean way to dig out the name of the offending field right now. Neither [`BuildError`](https://github.com/smithy-lang/smithy-rs/blob/2f29209c1ecfd8a848dc0745dbce6b510c898d9a/rust-runtime/aws-smithy-types/src/error/operation.rs#L91) nor its inner [`BuildErrorKind`](https://github.com/smithy-lang/smithy-rs/blob/2f29209c1ecfd8a848dc0745dbce6b510c898d9a/rust-runtime/aws-smithy-types/src/error/operation.rs#L67) give public visibility to their members.

Did you have a particular use-case in mind for needing to programmatically retrieve the name of the invalid field, rather than showing the full error message?

Reply:
I see what you mean - I had thought `BuildErrorKind` was structured rather than just a string to enable programmatic handling, but it looks more like the main purpose is just to display a nice error message string. 

Looks like we're meeting the bar of the existing `smithy-rs` behavior so I think this is the right call for now. We can potentially revisit and improve as part of #598